src/clients/arclib/arcstat.1.in
.\" -*- nroff -*-
.TH ARCSTAT 1 "@DATE@" "NorduGrid ARC @VERSION@" "NorduGrid Users Manual"

.SH NAME

arcstat \- ARC Status

.SH DESCRIPTION

The
.B arcstat
command is used for obtaining the status of jobs that have been
submitted to grid enabled resources.

.SH SYNOPSIS

.B arcstat [options] [job ...]

.SH OPTIONS

.IP "\fB-a\fR, \fB--all\fR"
all jobs
.IP "\fB-j\fR, \fB--joblist\fR=\fIfilename\fR"
the file storing information about active jobs (default ~/.arc/jobs.xml)
.IP "\fB-i\fR, \fB--jobids-from-file\fR=\fIfilename\fR"
a file containing a list of jobIDs
.IP "\fB-c\fR, \fB--cluster\fR=\fIname\fR"
select one or more computing elements: \fIname\fR can be an alias for a single CE, a group of CEs or a URL
.IP "\fB-I\fR, \fB--infointerface\fR=\fIInterfaceName\fR"
the computing element specified by URL at the command line should be
queried using this information interface (possible options:
org.nordugrid.ldapng, org.nordugrid.ldapglue2,
org.nordugrid.wsrfglue2, org.ogf.glue.emies.resourceinfo)
.IP "\fB-r\fR, \fB--rejectmanagement\fR=\fIURL\fR"
skip jobs which are on a computing element with a given URL
.IP "\fB-s\fR, \fB--status\fR=\fIstatusstr\fR"
only select jobs whose status is statusstr
.IP "\fB-l\fR, \fB--long\fR"
long format (more information)
.IP "\fB-S\fR, \fB--sort\fR"
sort jobs according to jobid, submissiontime or jobname
.IP "\fB-R\fR, \fB--rsort\fR"
reverse sorting of jobs according to jobid, submissiontime or jobname
.IP "\fB-p\fR, \fB--print-jobids\fR"
instead of the status only the IDs of the selected jobs will be printed
.IP "\fB-P\fR, \fB--listplugins\fR"
list the available plugins
.IP "\fB-t\fR, \fB--timeout\fR=\fIseconds\fR"
timeout in seconds (default 20)
.IP "\fB-z\fR, \fB--conffile\fR=\fIfilename\fR"
configuration file (default ~/.arc/client.conf)
.IP "\fB-d\fR, \fB--debug\fR=\fIdebuglevel\fR"
FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG
.IP "\fB-v\fR, \fB--version\fR"
print version information
.IP "\fB-?\fR, \fB--help\fR"
print help
.LP

.SH ARGUMENTS

.IP "\fBjob\fR ..."
list of jobids and/or jobnames
.LP

.SH EXTENDED DESCRIPTION

The
.B arcstat
command gives the status of a job submitted to a ARC enabled
resource. The job can be referred to either by the jobid that was
returned by
.BR arcsub (1)
at submission time or by its jobname if the job description that was
submitted contained a jobname attribute.

More than one jobid and/or jobname can be given. If several jobs were
submitted with the same jobname the status of all those jobs are
shown. If the
.B --joblist
option is used the list of jobs is read from a file with the
specified filename. By specifying the
.B --all
option, the status of all active jobs will be shown. If the
.B --long
option is given more detailed information is given.

Jobs can be sorted according to the jobid, submissiontime or jobname, either
in normal or reverse order. By using the
.B --sort
or
.B --rsort
option
followed by the desired ordering ('jobid', 'submissiontime' or 'jobname'), jobs
will be sorted in normal or reverse order. Note that the options
.B --sort
and
.B --rsort
cannot be used at the same time.

The
.B --cluster
option can be used to select or reject jobs at specific clusters. See
.BR arcsub (1)
for a discussion of the format of arguments to this option. The
.B --status
option can be used to select jobs in a specific state. These options
can be repeated several times.

The
.B arcstat
command locates the available clusters by querying information index
servers. Which servers to query can be specified by giving the
.B --index
option to the command. See
.BR arcsub (1)
for a discussion of the format of arguments to this option.

Also in this case the
.B --long
option can be used to obtain more detailed information.

Previously the
.B arcstat
command was also used to query information of clusters and/or index servers.
This functionality have been moved to the new command
.B arcinfo
.

.SH FILES

.TP
.B ~/.arc/client.conf
Some options can be given default values by specifying them in the
ARC client configuration file. By using the
.B --conffile
option a different configuration file can be used than the default.

.TP
.B ~/.arc/jobs.xml
This a local list of the user's active jobs. When a job is
successfully submitted it is added to this list and when it is removed
from the remote cluster it is removed from this list. This list is
used as the list of all active jobs when the user specifies the
.B --all
option to the various NorduGrid ARC user interface commands. By using the
.B --joblist
option a different file can be used than the default.

.SH ENVIRONMENT VARIABLES

.TP
.B X509_USER_PROXY
The location of the user's grid proxy file. Shouldn't be set unless
the proxy is in a non-standard location.

.TP
.B ARC_LOCATION
The location where ARC is installed can be specified by this variable. If not
specified the install location will be determined from the path to the
command being executed, and if this fails a WARNING will be given stating the
location which will be used.

.TP
.B ARC_PLUGIN_PATH
The location of ARC plugins can be specified by this variable. Multiple
locations can be specified by separating them by : (; in Windows). The
default location is \fB$ARC_LOCATION\fR/lib/arc (\\ in Windows).

.SH COPYRIGHT

APACHE LICENSE Version 2.0

.SH AUTHOR

Mattias Ellert <mattias.ellert@fysast.uu.se>

.SH SEE ALSO

.BR arccat (1),
.BR arcclean (1),
.BR arccp (1),
.BR arcget (1),
.BR arcinfo (1),
.BR arckill (1),
.BR arcls (1),
.BR arcmigrate (1),
.BR arcmkdir (1),
.BR arcproxy (1),
.BR arcrenew (1),
.BR arcresub (1),
.BR arcresume (1),
.BR arcrm (1),
.BR arcsub (1),
.BR arcsync (1),
.BR arctest (1)



NEWS
no news good news :)


src/clients/compute/arcstat.cpp
// -*- indent-tabs-mode: nil -*-

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <list>
#include <string>
#include <algorithm>

#include <arc/ArcLocation.h>
#include <arc/IString.h>
#include <arc/Logger.h>
#include <arc/UserConfig.h>
#include <arc/compute/JobSupervisor.h>

#include "utils.h"

int RUNMAIN(arcstat)(int argc, char **argv) {

  setlocale(LC_ALL, "");

  Arc::Logger logger(Arc::Logger::getRootLogger(), "arcstat");
  Arc::LogStream logcerr(std::cerr);
  logcerr.setFormat(Arc::ShortFormat);
  Arc::Logger::getRootLogger().addDestination(logcerr);
  Arc::Logger::getRootLogger().setThreshold(Arc::WARNING);

  Arc::ArcLocation::Init(argv[0]);

  ClientOptions opt(ClientOptions::CO_STAT,
                    istring("[job ...]"),
                    istring("The arcstat command is used for "
                            "obtaining the status of jobs that have\n"
                            "been submitted to Grid enabled resources."));

  std::list<std::string> jobidentifiers = opt.Parse(argc, argv);

  if (opt.showversion) {
    std::cout << Arc::IString("%s version %s", "arcstat", VERSION)
              << std::endl;
    return 0;
  }

  // If debug is specified as argument, it should be set before loading the configuration.
  if (!opt.debug.empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(opt.debug));

  if (opt.show_plugins) {
    std::list<std::string> types;
    types.push_back("HED:JobControllerPlugin");
    showplugins("arcstat", types, logger);
    return 0;
  }

  Arc::UserConfig usercfg(opt.conffile, opt.joblist);
  if (!usercfg) {
    logger.msg(Arc::ERROR, "Failed configuration initialization");
    return 1;
  }

  if (opt.debug.empty() && !usercfg.Verbosity().empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(usercfg.Verbosity()));

  for (std::list<std::string>::const_iterator it = opt.jobidinfiles.begin(); it != opt.jobidinfiles.end(); it++) {
    if (!Arc::Job::ReadJobIDsFromFile(*it, jobidentifiers)) {
      logger.msg(Arc::WARNING, "Cannot read specified jobid file: %s", *it);
    }
  }

  if (opt.timeout > 0)
    usercfg.Timeout(opt.timeout);

  if (!opt.sort.empty() && !opt.rsort.empty()) {
    logger.msg(Arc::ERROR, "The 'sort' and 'rsort' flags cannot be specified at the same time.");
    return 1;
  }

  if (!opt.rsort.empty()) {
    opt.sort = opt.rsort;
  }

  typedef bool (*JobSorting)(const Arc::Job&, const Arc::Job&);
  std::map<std::string, JobSorting> orderings;
  orderings["jobid"] = &Arc::Job::CompareJobID;
  orderings["submissiontime"] = &Arc::Job::CompareSubmissionTime;
  orderings["jobname"] = &Arc::Job::CompareJobName;

  if (!opt.sort.empty() && orderings.find(opt.sort) == orderings.end()) {
    std::cerr << "Jobs cannot be sorted by \"" << opt.sort << "\", the following orderings are supported:" << std::endl;
    for (std::map<std::string, JobSorting>::const_iterator it = orderings.begin();
         it != orderings.end(); it++)
      std::cerr << it->first << std::endl;
    return 1;
  }

  if ((!opt.joblist.empty() || !opt.status.empty()) && jobidentifiers.empty() && opt.clusters.empty())
    opt.all = true;

  if (jobidentifiers.empty() && opt.clusters.empty() && !opt.all) {
    logger.msg(Arc::ERROR, "No jobs given");
    return 1;
  }
  
  std::list<std::string> selectedURLs;
  if (!opt.clusters.empty()) {
    selectedURLs = getSelectedURLsFromUserConfigAndCommandLine(usercfg, opt.clusters);
  }
  std::list<std::string> rejectManagementURLs = getRejectManagementURLsFromUserConfigAndCommandLine(usercfg, opt.rejectmanagement);
  std::list<Arc::Job> jobs;
  if (!Arc::Job::ReadJobsFromFile(usercfg.JobListFile(), jobs, jobidentifiers, opt.all, selectedURLs, rejectManagementURLs)) {
    logger.msg(Arc::ERROR, "Unable to read job information from file (%s)", usercfg.JobListFile());
    return 1;
  }

  for (std::list<std::string>::const_iterator itJIDAndName = jobidentifiers.begin();
       itJIDAndName != jobidentifiers.end(); ++itJIDAndName) {
    std::cout << Arc::IString("Warning: Job not found in job list: %s", *itJIDAndName) << std::endl;
  }

  Arc::JobSupervisor jobmaster(usercfg, jobs);
  jobmaster.Update();
  jobmaster.SelectValid();
  if (!opt.status.empty()) {
    jobmaster.SelectByStatus(opt.status);
  }
  jobs = jobmaster.GetSelectedJobs();

  if (jobs.empty()) {
    std::cout << Arc::IString("No jobs") << std::endl;
    return 1;
  }

  std::vector<Arc::Job> jobsSortable(jobs.begin(), jobs.end());

  if (!opt.sort.empty()) {
    opt.rsort.empty() ? std::sort(jobsSortable.begin(),  jobsSortable.end(),  orderings[opt.sort]) :
                        std::sort(jobsSortable.rbegin(), jobsSortable.rend(), orderings[opt.sort]);
  }

  for (std::vector<Arc::Job>::const_iterator it = jobsSortable.begin();
       it != jobsSortable.end(); it++) {
    // Option 'long' (longlist) takes precedence over option 'print-jobids' (printids)
    if (opt.longlist || !opt.printids) {
      it->SaveToStream(std::cout, opt.longlist);
    }
    else {
      std::cout << it->JobID << std::endl;
    }
  }

  return 0;
}
